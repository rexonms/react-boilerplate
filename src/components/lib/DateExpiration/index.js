import React from 'react';
import PropTypes from 'prop-types';

import Styles from './DateExpiration.scss';

const propTypes = {
  /** Epoch timestamp */
  timestamp: PropTypes.number.isRequired,
  /** Min num of days that the message should display */
  minDay: PropTypes.number,
  /** String that displays expired */
  expired: PropTypes.string.isRequired,
  /** String that displays expired in n days */
  expiresIn: PropTypes.shape({
    singular: PropTypes.string.isRequired,
    plural: PropTypes.string.isRequired,
  }).isRequired,
};

const defaultProps = {
  minDay: 30,
};

export const epochDay = 60 * 60 * 24;
export const currentTimeStamp = Math.floor(Date.now() / 1000);

const getText = (text, day) => text.replace('#NUM#', day);

export const timestampExpiresIn = (expiringTimestamp, minDay) => {
  const remainingDays = (expiringTimestamp - currentTimeStamp) / epochDay;
  // Already expired
  if (expiringTimestamp < currentTimeStamp) {
    return -1;
  }
  // Show remaining days
  if (remainingDays < minDay) {
    return remainingDays;
  }
  return 0;
};

const DateExpiration = (props) => {
  const expiresInDay = timestampExpiresIn(props.timestamp, props.minDay);
  if (expiresInDay === 0) {
    return null;
  }

  if (expiresInDay === -1) {
    return (
      <span className={`${Styles.container}`}>
        <p>{props.expired}</p>
      </span>
    );
  }
  return (
    <span className={`${Styles.container}`}>
      <p>
        {expiresInDay === 1 ?
          getText(props.expiresIn.singular, expiresInDay) :
          getText(props.expiresIn.plural, expiresInDay)}
      </p>
    </span>
  );
};

DateExpiration.propTypes = propTypes;
DateExpiration.defaultProps = defaultProps;

export default DateExpiration;
