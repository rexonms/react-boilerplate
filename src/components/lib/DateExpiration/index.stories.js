import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number } from '@storybook/addon-knobs/react';
import { withInfo } from '@storybook/addon-info';

import DateExpiration, { epochDay, currentTimeStamp } from './index';

const stories = storiesOf('components/lib/DateExpiration', module);
export const props = {
  timestamp: 1527633993,
  minDay: 30,
  expired: '(expired)',
  expiresIn: {
    singular: '(expires in #NUM# day)',
    plural: '(expires in #NUM# days)',
  },
};

stories.addDecorator(withKnobs);
stories.addDecorator(story => (story()));
stories.add('DateExpiration', withInfo(`
  Display a message if the epoch date will expire with in n days or expired. If the date is > n days it will return null
  ~~~js
    <DateExpiration name='DateExpiration' />
  ~~~
`)(() => {
  const label = 'timestamp';
  const defaultValue = currentTimeStamp + (epochDay * 5);
  const options = {
    range: true,
    min: currentTimeStamp - epochDay,
    max: currentTimeStamp + (epochDay * 40),
    step: epochDay,
  };
  const groupId = 'GROUP-ID1';
  const timestampOptions = number(label, defaultValue, options, groupId);

  return (
    <DateExpiration
      timestamp={timestampOptions}
      expired={props.expired}
      expiresIn={props.expiresIn}
    />);
}));
