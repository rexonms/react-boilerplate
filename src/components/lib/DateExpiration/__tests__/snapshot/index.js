import React from 'react';
import renderer from 'react-test-renderer';

import DateExpiration from '../../index';
import { props } from './../../index.stories';

describe('DateExpiration Component', () => {
  it('Snapshot matches', () => {
    const tree = renderer.create(<DateExpiration {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});