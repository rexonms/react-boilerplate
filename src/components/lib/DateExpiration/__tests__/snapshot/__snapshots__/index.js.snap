// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateExpiration Component Snapshot matches 1`] = `
<span
  className="container"
>
  <p>
    (expired)
  </p>
</span>
`;
